main/src/ca/uwaterloo/flix/language/phase/OccurrenceAnalyzer1.scala

```diff
@@ -0,0 +1,580 @@
+        val o2 = o1 + (sym -> Once)
+        (OccurrenceAst1.Expr.ApplyLocalDef(sym, es, tpe, eff, loc), increment(o2))
+
+      case MonoAst.Expr.Let(sym, exp1, exp2, tpe, eff, loc) =>
+        val (e1, o1) = visit(exp1)(Some(sym))
+        val (e2, o2) = visit(exp2)
+        val o3 = combineInfo(o1, o2)
+        val occur = o3.get(sym)
+        val o4 = o3 - sym
+        (OccurrenceAst1.Expr.Let(sym, e1, e2, tpe, eff, occur, loc), increment(o4))
+
+      case MonoAst.Expr.LocalDef(sym, fparams, exp1, exp2, tpe, eff, loc) =>
+        val fps = fparams.map(visitFormalParam)
+        val (e1, o10) = visit(exp1)
+        val o1 = captureVars(o10)
+        val (e2, o2) = visit(exp2)
+        val o3 = combineInfo(o1, o2)
+        val occur = o3.get(sym)
+        val o4 = o3 - sym
+        (OccurrenceAst1.Expr.LocalDef(sym, fps, e1, e2, tpe, eff, occur, loc), increment(o4))
+
+      case MonoAst.Expr.Scope(sym, rvar, exp, tpe, eff, loc) =>
+        val (e, o1) = visit(exp)
+        val o2 = o1 :+ sym0 -> DontInline
+        (OccurrenceAst1.Expr.Scope(sym, rvar, e, tpe, eff, loc), increment(o2))
+
+      case MonoAst.Expr.IfThenElse(exp1, exp2, exp3, tpe, eff, loc) =>
+        val (e1, o1) = visit(exp1)
+        val (e2, o2) = visit(exp2)
+        val (e3, o3) = visit(exp3)
+        val o4 = combineInfo(o1, combineInfoBranch(o2, o3))
+        (OccurrenceAst1.Expr.IfThenElse(e1, e2, e3, tpe, eff, loc), increment(o4))
+
+      case MonoAst.Expr.Stm(exp1, exp2, tpe, eff, loc) =>
+        val (e1, o1) = visit(exp1)
+        val (e2, o2) = visit(exp2)
+        val o3 = combineInfo(o1, o2)
+        (OccurrenceAst1.Expr.Stm(e1, e2, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.Discard(exp, eff, loc) =>
+        val (e, o) = visit(exp)
+        (OccurrenceAst1.Expr.Discard(e, eff, loc), increment(o))
+
+      case MonoAst.Expr.Match(exp, rules, tpe, eff, loc) =>
+        val (e, o1) = visit(exp)
+        val (rs, o2) = visitMatchRules(rules)
+        val o3 = combineInfo(o1, o2)
+        (OccurrenceAst1.Expr.Match(e, rs, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.VectorLit(exps, tpe, eff, loc) =>
+        val (es, o) = visitExps(exps)
+        (OccurrenceAst1.Expr.VectorLit(es, tpe, eff, loc), increment(o))
+
+      case MonoAst.Expr.VectorLoad(exp1, exp2, tpe, eff, loc) =>
+        val (e1, o1) = visit(exp1)
+        val (e2, o2) = visit(exp2)
+        val o3 = combineInfo(o1, o2)
+        (OccurrenceAst1.Expr.VectorLoad(e1, e2, tpe, eff, loc), increment(o3))
+
+      case MonoAst.Expr.VectorLength(exp, loc) =>
+        val (e, o) = visit(exp)
+        (OccurrenceAst1.Expr.VectorLength(e, loc), increment(o))
+
+      case MonoAst.Expr.Ascribe(exp, tpe, eff, loc) =>
+        val (e, o) = visit(exp)
+        (OccurrenceAst1.Expr.Ascribe(e, tpe, eff, loc), increment(o))
+
+      case MonoAst.Expr.Cast(exp, declaredType, declaredEff, tpe, eff, loc) =>
+        val (e, o) = visit(exp)
+        val o1 = o :+ sym0 -> Dangerous
```

Comment:
This could be more finegrained, like for `def example(): Unit = (_ -> ???)()` it is not example that is dangerous but the lambda. same with local def

Reply:
also if decalredEff is None then i don't think you have to put dangerous